### ISSUE TITLE
Crash on macos catalina: UnboundLocalError: local variable 'line' referenced before assignment

### ISSUE THREAD

**Issue by kim0** [dlenski/vpn-slice#67]
Hi again,

I usually am connected to my personal wireguard VPN, then use openconnect to connect to work VPN. This works well after you last fix. However today I tried connecting to work, without being connected to wireguard first, this caused the crash below! I tested this twice and then connected to wireguard, then tested twice and it's very consistent. Hope this is enough to fix the below crash.

```
WARNING: no firewall provider available; can't block incoming traffic
route: writing to routing socket: not in table
Traceback (most recent call last):
  File "/usr/local/bin/vpn-slice", line 11, in <module>
    load_entry_point('vpn-slice==0.14.2', 'console_scripts', 'vpn-slice')()
  File "/usr/local/Cellar/vpn-slice/0.14.2/libexec/lib/python3.8/site-packages/vpn_slice/__main__.py", line 520, in main
    do_connect(env, args)
  File "/usr/local/Cellar/vpn-slice/0.14.2/libexec/lib/python3.8/site-packages/vpn_slice/__main__.py", line 176, in do_connect
    exc_subnets = [(dest, providers.route.get_route(dest)) for dest in args.exc_subnets]
  File "/usr/local/Cellar/vpn-slice/0.14.2/libexec/lib/python3.8/site-packages/vpn_slice/__main__.py", line 176, in <listcomp>
    exc_subnets = [(dest, providers.route.get_route(dest)) for dest in args.exc_subnets]
  File "/usr/local/Cellar/vpn-slice/0.14.2/libexec/lib/python3.8/site-packages/vpn_slice/mac.py", line 70, in get_route
    keys = line.split()
UnboundLocalError: local variable 'line' referenced before assignment
Script 'vpn-slice -v -S' returned error 1
```

**Comment by dlenski** [comment]
Standard disclaimer: I don't use macOS, can't test this, and didn't write this part of the code (ping @gmacon who did :sunglasses:)… 

It appears that the `route get` command is unexpectedly not returning anything for this query:

https://github.com/dlenski/vpn-slice/blob/5cd3286df133166b9a357cd2673d77fc17cdbb78/vpn_slice/mac.py#L65-L69

Can you run `route -n get -inet IP_ADDRESS_OF_VPN_GATEWAY` and show the output (fine to obscure the IP address)? It _should_ look like [this](https://unix.stackexchange.com/questions/53446/what-is-the-gnu-linux-equivalent-of-bsds-route-get), but apparently doesn't in this case.

**Comment by gmacon** [comment]
It looks like you asked the right questions. I don't have any immediate hypotheses as to why `route` wouldn't return anything.

This is why you're not supposed to use the iteration variable outside of the loop body, oops.

Your fix looks correct to me, at least to the extent that it will give a helpful error instead of an unhelpful one.

**Comment by kim0** [comment]
`route get 62.134.a.b` seems to work normally.

```bash
   route to: 62.134.a.b
destination: 62.134.a.b
    gateway: 192.168.33.1
  interface: en0
      flags: <UP,GATEWAY,HOST,DONE,WASCLONED,IFSCOPE,IFREF>
 recvpipe  sendpipe  ssthresh  rtt,msec    rttvar  hopcount      mtu     expire
       0         0         0         0         0         0      1500         0
```

After applying your fix, I get the below crash

```
  CISCO_*SPLIT_INC_*      => splitinc=[IPv4Network('192.168.0.0/16'), IPv4Network('10.0.0.0/8')]
  CISCO_*SPLIT_EXC_*      => splitexc=[IPv4Network('0.0.0.0/32')]
WARNING: no firewall provider available; can't block incoming traffic
route: writing to routing socket: not in table
Traceback (most recent call last):
  File "/usr/local/bin/vpn-slice", line 11, in <module>
    load_entry_point('vpn-slice==0.14.2', 'console_scripts', 'vpn-slice')()
  File "/usr/local/Cellar/vpn-slice/0.14.2/libexec/lib/python3.8/site-packages/vpn_slice/__main__.py", line 520, in main
    do_connect(env, args)
  File "/usr/local/Cellar/vpn-slice/0.14.2/libexec/lib/python3.8/site-packages/vpn_slice/__main__.py", line 176, in do_connect
    exc_subnets = [(dest, providers.route.get_route(dest)) for dest in args.exc_subnets]
  File "/usr/local/Cellar/vpn-slice/0.14.2/libexec/lib/python3.8/site-packages/vpn_slice/__main__.py", line 176, in <listcomp>
    exc_subnets = [(dest, providers.route.get_route(dest)) for dest in args.exc_subnets]
  File "/usr/local/Cellar/vpn-slice/0.14.2/libexec/lib/python3.8/site-packages/vpn_slice/mac.py", line 75, in get_route
    raise OSError("'route get {} {}' returned neither 'interface' nor 'gateway' fields".format(
OSError: 'route get -inet 0.0.0.0/32' returned neither 'interface' nor 'gateway' fields
Script 'vpn-slice -v -S --no-ns-hosts -D' returned error 1
```

The all zeros IP, seems to be pushed from the server as an "Exclude" .. not sure why it's being looked up with route.

**Comment by gmacon** [comment]
Excluded networks are looked up with `route` to figure out what the gateway is before we set up the new routes. That makes sure that the excluded traffic routes to the same place as it did before even if it didn't previously route out the default gateway.

Excluding 0.0.0.0/32 seems a little odd to me (there should never be any traffic addressed *to* that address), but I was able to reproduce the no-output-from-route issue on my Mac; it happens when there is no route to a particular address. (I got really lucky on that reproduction, because my laptop happened to be slow connecting to the WiFi this morning and I managed to run `route get -inet 0.0.0.0/32` before the default route got set up.) We might be able to demote this error to a warning along the lines of

    WARNING: No route to excluded network 0.0.0.0/32, not restoring a route.

**Comment by dlenski** [comment]
Aha! The change is doing exactly what we need… showing where the problem actually lies.

This issue is closely related to [OpenConnect gitlab issue #172](https://gitlab.com/openconnect/openconnect/-/issues/172#note_394198461) (which really should be changed to a vpnc-script issue).

Basically, we need to ignore `0.0.0.0/x` as a split-exclude. Adding a patch to warn about that. (Thanks @gmacon, that language is useful.)

**Comment by gmacon** [comment]
We should probably ignore all of the [special-purpose blocks](https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml) with Destination=False. [IPv6 version](https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml)

By my reckoning, that's:

```
0.0.0.0/8
192.0.0.8/32
192.0.0.170/31
192.0.2.0/24
198.51.100.0/24
203.0.113.0/24
240.0.0.0/4
127.0.0.0/8
::1/128
::/128
::ffff:0:0/96
2001:db8::/32
```

I've skipped the "IETF Protocol Assignments" blocks because there are more specific addresses within those blocks with Destination=True.

**Comment by dlenski** [comment]
> We should probably ignore all of the [special-purpose blocks](https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml) with Destination=False. [IPv6 version](https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml)

Thanks @gmacon. I wish that the `ipaddress` module had an [`is_*` method](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address.is_multicast) that matched all of these.
I can't think of any reason why a VPN should ever touch routing to these in any way… anyone else? Then again, there are lots of other things a VPN _probably should not_ override, but might. :thinking:

For the moment I think I'm satisfied with simply converting errors in `.get_route`/`.add_route`/`.warn_route` to explicit warnings.

**Comment by dlenski** [comment]
I made the Linux implementation (`Iproute2Provider`) return `None` for localhost routes, since trying to add/replace these causes errors… should the same be done for `BSDRouteProvider`?

Per https://www.pprune.org/archive/index.php/t-441233.html, it appears the output of `route -n get 127.0.0.1` is not distinguishable by either interface name (is that always `lo0`?) or by the `LOCAL` flag… do I have that about route?

```
route to: 127.0.0.1
destination: 127.0.0.1
interface: lo0
flags: <UP,HOST,DONE,LOCAL>
```

**Comment by kim0** [comment]
@dlenski Thanks a lot! Can you please cut a release, thanks!

**Comment by dlenski** [comment]
> @dlenski Thanks a lot! Can you please cut a release, thanks!

Did you test. https://github.com/dlenski/vpn-slice/commit/270a0c25126fa7df9003a40e7bf951aa9b0fc17e

**Comment by kim0** [comment]
@dlenski .. Yes I tested that fix and it is able to connect, thanks!

**Comment by dlenski** [comment]
:+1: Released as [v0.15](https://github.com/dlenski/vpn-slice/releases/tag/v0.15).
